File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
class Solution {

    public boolean check_dist(String[] place, int x, int y) {
        int[][] oneDir = {{1,0},{0,1}};
        int[][] twoDir = {{2,0},{0,2}};
        int[][] crossDir = {{1,1},{1,-1}};
        
        for(int[] dir: oneDir) {
            int nx = x+dir[0], ny = y+dir[1];
            if(0<=nx && nx<5 && 0<=ny && ny<5) {
                if (place[nx].charAt(ny) == 'P') return false;
            }
        }
        
        for(int[] dir: crossDir) {
            int nx = x+dir[0], ny = y+dir[1];
            if(0<=nx && nx<5 && 0<=ny && ny<5 && place[nx].charAt(ny) == 'P') {
                if (!(place[nx].charAt(y) == 'X' && place[x].charAt(ny) == 'X')) {
                    return false;
                }
            }
        }
        
        int tmp = 0;
        for(int[] dir: twoDir) {
            int nx = x+dir[0], ny = y+dir[1];
            if(0<=nx && nx<5 && 0<=ny && ny<5 && place[nx].charAt(ny) == 'P') {
                if(place[x+oneDir[tmp][0]].charAt(y+oneDir[tmp][1]) != 'X') return false;
            }
            tmp++;
        }
        
        return true;
    }    
    
    public int[] solution(String[][] places) {
        int[] answer = new int[5];
        
        for(int p=0;p<5;p++) {
            boolean flag = true;
            for(int i=0;i<5;i++) {
                int j = 0;
                for(char c: places[p][i].toCharArray()) {
                    if (c == 'P') {
                        if (!check_dist(places[p], i, j)) {
                            flag = false;
                            break;
                        }
                    }
                    j++;
                }
                if (!flag) break;
            }
            answer[p] = flag ? 1 : 0;
        }
        
        return answer;
    }
}

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.swift
import Foundation

func solution(_ places:[[String]]) -> [Int] {
    
    let n = 5
    var result = [Int]()
    
    func check(
        _ sx: Int, 
        _ sy: Int, 
        _ grid: [[Character]]
    ) -> Bool {
        let oneDir = [(1,0),(-1,0),(0,1),(0,-1)]
        let twoDir = [(0,2),(2,0)]
        let crossDir = [(1,1),(1,-1)]
        
        for d in oneDir {
            let nx = sx+d.0, ny = sy+d.1
            if(0>nx || nx>=n || 0>ny || ny>=n) { continue }
            if grid[nx][ny] == "P" { return false }
        }
        
        for d in twoDir {
            let nx = sx+d.0, ny = sy+d.1
            if(0>nx || nx>=n || 0>ny || ny>=n) { continue }
            if grid[nx][ny] == "P" && grid[sx+d.0/2][sy+d.1/2] != "X" { 
                return false
            }
        }
        
        for d in crossDir {
            let nx = sx+d.0, ny = sy+d.1
            if(0>nx || nx>=n || 0>ny || ny>=n) { continue }
            if grid[nx][ny] == "P" { 
                if grid[sx][ny] != "X" || grid[nx][sy] != "X" {
                    return false
                }
            }
        }
        
        return true
    }
    
    for p in places {
        let place = p.map {Array($0)}
        var ans = true
        for i in 0..<n {
            for j in 0..<n {
                if place[i][j] == "P" {
                    if !check(i,j,place) {
                        ans = false
                        break
                    }
                }
            }
            if !ans { break }
        }
        result.append(ans ? 1 : 0)
    }

    
    
    return result
}